Feature actuator disk with variable load

EttoreSaetta opened the pull request:
## Proposed Changes
Addition of a new actuator disk model with a variable load and swirl distributions along disk radius.

Contributors:
Theoretical and Applied Aerodynamic Research Group (TAARG) of University of Naples Federico II.
EttoreSaetta, lorenzorusso07, rtogna.

## Related Work
Introduction of new options in the config. file.
Introduction of an additional input file that contains the propeller data.
Introduction of additional c++ functions.
Addition of code lines to existing functions preserving the original actuator disk model.

## PR Checklist
- [X ] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [ X] My contribution is commented and consistent with SU2 style.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

juanjosealonso commented:
Ettore,

Thanks for submitting this.  I thought you gave an excellent presentation at the SU2 Conference and the much better actuator disk model you and your colleagues have put together will surely be useful to many people.

Thanks for participating and being a member of the SU2 Community.  Please stay healthy and well and I hope to meet you in person next time we have a meeting (fingers crossed….maybe next year?).

Best wishes,

Juan


> On Jun 18, 2020, at 1:27 AM, SaettaE <notifications@github.com> wrote:
> 
> Proposed Changes
> 
> Addition of a new actuator disk model with a variable load and swirl distributions along disk radius.
> 
> Contributors:
> Theoretical and Applied Aerodynamic Research Group (TAARG) of University of Naples Federico II.
> EttoreSaetta, lorenzorusso07, rtogna.
> 
> Related Work
> 
> Introduction of new options in the config. file.
> Introduction of an additional input file that contains the propeller data.
> Introduction of additional c++ functions.
> Addition of code lines to existing functions preserving the original actuator disk model.
> 
> PR Checklist
> 
> [X ] I am submitting my contribution to the develop branch.
>  My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
> [ X] My contribution is commented and consistent with SU2 style.
>  I have added a test case that demonstrates my contribution, if necessary.
>  I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/su2code/SU2/pull/1027 <https://github.com/su2code/SU2/pull/1027>
> Commit Summary
> 
> Update config_template.cfg
> File Changes
> 
> M config_template.cfg <https://github.com/su2code/SU2/pull/1027/files#diff-2c2f226c0183e7633cf165c6d3862450> (5)
> Patch Links:
> 
> https://github.com/su2code/SU2/pull/1027.patch <https://github.com/su2code/SU2/pull/1027.patch>
> https://github.com/su2code/SU2/pull/1027.diff <https://github.com/su2code/SU2/pull/1027.diff>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/su2code/SU2/pull/1027>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA5FFRGLBAHYBVTV3GBSLATRXHFYXANCNFSM4OBLB3CQ>.
>

EttoreSaetta commented:
> Ettore, Thanks for submitting this. I thought you gave an excellent presentation at the SU2 Conference and the much better actuator disk model you and your colleagues have put together will surely be useful to many people. Thanks for participating and being a member of the SU2 Community. Please stay healthy and well and I hope to meet you in person next time we have a meeting (fingers crossed….maybe next year?). Best wishes, Juan
> […](#)
> On Jun 18, 2020, at 1:27 AM, SaettaE ***@***.***> wrote: Proposed Changes Addition of a new actuator disk model with a variable load and swirl distributions along disk radius. Contributors: Theoretical and Applied Aerodynamic Research Group (TAARG) of University of Naples Federico II. EttoreSaetta, lorenzorusso07, rtogna. Related Work Introduction of new options in the config. file. Introduction of an additional input file that contains the propeller data. Introduction of additional c++ functions. Addition of code lines to existing functions preserving the original actuator disk model. PR Checklist [X ] I am submitting my contribution to the develop branch. My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags). [ X] My contribution is commented and consistent with SU2 style. I have added a test case that demonstrates my contribution, if necessary. I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary. You can view, comment on, or merge this pull request online at: #1027 <#1027> Commit Summary Update config_template.cfg File Changes M config_template.cfg <https://github.com/su2code/SU2/pull/1027/files#diff-2c2f226c0183e7633cf165c6d3862450> (5) Patch Links: https://github.com/su2code/SU2/pull/1027.patch <https://github.com/su2code/SU2/pull/1027.patch> https://github.com/su2code/SU2/pull/1027.diff <https://github.com/su2code/SU2/pull/1027.diff> — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <#1027>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA5FFRGLBAHYBVTV3GBSLATRXHFYXANCNFSM4OBLB3CQ>.

Thank you very much, I hope we can give an useful contribution.
I hope to see you in person too.

Best regards,
Ettore

EttoreSaetta commented:
> Hello @EttoreSaetta,
> I'm glad we got a pull request out of the meeting, especially one so well written like this one, very tidy code.
> I leave a few comments below in case you want to polish it up even more.
> Cheers,
> Pedro

Hello @pcarruscag,
thank you for reviewing our code and for all the valuable advice.
We are going to make the changes you suggest as soon as possible.

Have a nice day,
Ettore

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,194 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+  su2double Dens_FreeStream = config->GetDensity_FreeStream();
+  su2double *Vel_FreeStream = config->GetVelocity_FreeStream();
+
+  /*--- Variable load distribution is in input file. ---*/
+  /*--- Input file provides force coefficients distributions along disk radius. Initialization
```
Ah just noticed this block of code is in SetActDisk_BCThrust, can it be moved to its own private member function? We should try to keep functions on the smallish side (e.g. fits in one screen).

EttoreSaetta replied:
Did it. I moved it in another function named `ReadActDisk_InputFile`. Now the `SetActDisk_BCThrust` is more readable.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,194 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+                    for (iEl = 1; iEl < nRow; iEl++){
+                      Fa[iEl] = (dCt_v[iEl]*(2*Dens_FreeStream*Vel_FreeStream[0]*Vel_FreeStream[0])/
+                                (AD_J*AD_J*PI_NUMBER*rad_v[iEl])) / config->GetPressure_Ref();
+                      Ft[iEl] = (dCp_v[iEl]*(2*Dens_FreeStream*Vel_FreeStream[0]*Vel_FreeStream[0])/
```
Done and tested. Thank you.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -11097,6 +11392,251 @@ void CEulerSolver::BC_ActDisk(CGeometry *geometry, CSolver **solver_container, C
+  su2double Pressure_out, Density_out,
+  Pressure_in, Density_in;
+
+  su2double C[3], Prop_Axis[3], R, r[3], r_;
```
Done and tested. Thank you.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -11097,6 +11392,251 @@ void CEulerSolver::BC_ActDisk(CGeometry *geometry, CSolver **solver_container, C
+  bool ratio              = (config->GetActDisk_Jump() == RATIO);
+
+  su2double *Normal = new su2double[nDim];
+
```
Done and tested. Thank you.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -11097,6 +11392,251 @@ void CEulerSolver::BC_ActDisk(CGeometry *geometry, CSolver **solver_container, C
+      FQ = Q_out/Density_Disk;
+
+      /*--- Computation of the momentum jumps due to the tnagential and radial forces per unit area. ---*/
+      if (FQ < 1e-15){
```
Done and tested. Thank you.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,194 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
   Ref = config->GetDensity_Ref() * config->GetVelocity_Ref() * config->GetVelocity_Ref() * 1.0 * 1.0;
 
+  su2double Dens_FreeStream = config->GetDensity_FreeStream();
+  su2double *Vel_FreeStream = config->GetVelocity_FreeStream();
```
Did it. Thank you.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,194 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+
+           if (!ActDisk_file.fail()) {
+              string text_line, text_line_appo, name[2];
+              string::size_type position;
```
Did it.
Great quote, I will keep it in mind.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,194 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+                          r[iDim] = P[iDim]-AD_Center[iDim];
+                          r_ += r[iDim]*r[iDim];
+                       }
+                       r_ = sqrt(r_);
```
Did it, I used the functions from the "GeometryToolbox".

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -11097,6 +11392,251 @@ void CEulerSolver::BC_ActDisk(CGeometry *geometry, CSolver **solver_container, C
+
+      Area = 0.0;
+      for (iDim = 0; iDim < nDim; iDim++) Area += Normal[iDim]*Normal[iDim];
+      Area = sqrt (Area);
```
Did it. Also here I used the GeometryToolbox.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -11097,6 +11392,251 @@ void CEulerSolver::BC_ActDisk(CGeometry *geometry, CSolver **solver_container, C
+          for (iDim = 0; iDim < nDim; iDim++) QdMnorm2 += QdMnorm[iDim]*QdMnorm[iDim];
+
+          /*--- Resolving the second grade equation for the density. ---*/
+          appo2 = -((2*sqrt(max(0.0,QdMnorm2))*RiemannExtr)+((4*Gamma*Pressure_out)/(Gamma_Minus_One*Gamma_Minus_One)));
```
We have notice that QdMnorm2 is never negative, so I have just removed the `max()` function.
I have used `pow(..., 2)` where necessary.

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,189 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+                     GeometryToolbox::Distance(nDim, P, AD_Center, r);
+
+                     /*--- Computation of the non-dimensional radius for the current node. ---*/
+                     r_ = GeometryToolbox::Distance(nDim, P, AD_Center) / AD_Radius;
```
You could use just `Norm` on the `r` computed above.

EttoreSaetta replied:
Did it.

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5764,9 +5860,189 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+                      Fr[0] = 0.0;
+                  }
+                  else{
+                    Fa[0] = (dCt_v[0]*(2*Dens_FreeStream*pow(Vel_FreeStream[0],2))/
+                            (pow(AD_J,2)*PI_NUMBER*rad_v[0])) / config->GetPressure_Ref();
```
:+1: thanks to me it reads more naturally like this.

EttoreSaetta replied:
Yes, it is more clear like this!

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -11097,6 +11387,249 @@ void CEulerSolver::BC_ActDisk(CGeometry *geometry, CSolver **solver_container, C
+  bool tkeNeeded          = (config->GetKind_Turb_Model() == SST) || (config->GetKind_Turb_Model() == SST_SUST);
+  bool ratio              = (config->GetActDisk_Jump() == RATIO);
+
+  su2double *Normal = new su2double[MAXNDIM];
```
Actually I meant `su2double Normal[MAXNDIM]` like most of your other variables, it's a minor detail.

EttoreSaetta replied:
Did it.

EttoreSaetta commented:
> Thank you for the changes LGTM, it would be nice if that section of code that does the reading could go to a separate function, unless it needs a lot of the information available where you have it now.
> 
> Let's try and get another review so that this PR can be merged before #1044.

Yes, we planned to put that piece of code into a separate function tomorrow. So I will try and push a commit for another review tomorrow.
Thank you again for all the advice.

EttoreSaetta commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -5745,6 +5834,13 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
   su2double MyBCThrust, BCThrust_Init;
   su2double Vector[MAXNDIM] = {0.0};
+  int iRow, nRow, iEl;
+  su2double rad_v[50]={0.0}, dCt_v[50]={0.0}, dCp_v[50]={0.0}, dCr_v[50]={0.0},
```
I have removed the "magic numbers": because of the fact that the dimension of these vectors depends on the actuator disk input file, I have decided to introduce a dynamic allocation, so that their dimension can change depending on the input file (which can contains different sizes if more than one actuator disk is considered).
I hope that this solution can be good enough.

pcarruscag commented on SU2_CFD/include/solvers/CSolver.hpp:
```diff
@@ -1101,6 +1101,24 @@ class CSolver {
+   * \param[in] config - Definition of the particular problem.
+   * \param[in] val_marker - Surface marker where the boundary condition is applied.
+   * \param[in] val_inlet_surface - Boolean for whether val_marker is an inlet
+   */
+  inline virtual void BC_ActDisk_VariableLoad(CGeometry *geometry,
+                                              CSolver **solver_container,
```
This function does not need to be virtual, since it is only used in CEulerSolver, by CEulerSolver

EttoreSaetta replied:
Did it, it is not anymore virtual.

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -306,6 +306,23 @@ CEulerSolver::CEulerSolver(CGeometry *geometry, CConfig *config,
+  ActDisk_Axis.resize(nMarker, MAXNDIM);
+
+  /*--- Actuator Disk Fa, Fx, Fy and Fz allocations ---*/
+  for (iMarker = 0; iMarker < nMarker; iMarker++) {
+    ActDisk_Fa.resize(nMarker, nVertex[iMarker]);
+    ActDisk_Fx.resize(nMarker, nVertex[iMarker]);
+    ActDisk_Fy.resize(nMarker, nVertex[iMarker]);
+    ActDisk_Fz.resize(nMarker, nVertex[iMarker]);
+  }
```
Sorry @EttoreSaetta, I think I gave you some unclear information, for the matrix type all rows need to have the same dimension, so for these 4 variables it would not work unfortunately.
For the other 3 variables it's fine, thanks for changing them.

EttoreSaetta replied:
Yes, I am sorry, I have noticed it, and I am already fixing it.

EttoreSaetta replied:
Did it, these 4 variables are not anymore `su2activematrix`, and I have restored the previous version for them.

pcarruscag commented on SU2_CFD/src/solvers/CEulerSolver.cpp:
```diff
@@ -6170,6 +6194,214 @@ void CEulerSolver::SetActDisk_BCThrust(CGeometry *geometry, CSolver **solver_con
+
+             /*--- Assign the vectors dimension. ---*/
+             rad_v = new su2double [nRow];
+             dCt_v = new su2double [nRow];
+             dCp_v = new su2double [nRow];
+             dCr_v = new su2double [nRow];
```
It does sound like the more robust way of doing it, but this code will likely leak memory, because the allocation is done inside the loop and the deallocation outside.
It would be safer (and avoid leaks) to use either std::vector or su2activevector for these variables that need to be resized dynamically.

EttoreSaetta replied:
Thank you for the advice, I have changed them as `std::vector`.

EttoreSaetta commented:
Thank you for the review @pcarruscag.
I merge the pull request.

pcarruscag commented:
Anytime.
